man/notify-selfcontained-example.c

```diff
@@ -0,0 +1,167 @@
+/* SPDX-License-Identifier: MIT-0 */
+
+/* Implement the systemd notify protocol without external dependencies.
+ * Supports both readiness notification on startup and on reloading,
+ * according to the protocol defined at:
+ * https://www.freedesktop.org/software/systemd/man/latest/sd_notify.html
+ * This protocol is guaranteed to be stable as per:
+ * https://systemd.io/PORTABILITY_AND_STABILITY/ */
+
+#include <assert.h>
+#include <errno.h>
+#include <inttypes.h>
+#include <signal.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <sys/socket.h>
+#include <sys/un.h>
+#include <time.h>
+#include <unistd.h>
+
+#define _cleanup_(f) __attribute__((cleanup(f)))
+
+bool reloading = 0;
+bool terminating = 0;
+
+static void closep(int *fd) {
+  if (!fd || *fd < 0)
+    return;
+
+  close(*fd);
+  *fd = -1;
+}
+
+static int notify(const char *message) {
+  size_t path_length, message_length;
+  struct sockaddr_un socket_addr = {
+    .sun_family = AF_UNIX,
+  };
+  _cleanup_(closep) int fd = -1;
+  const char *socket_path;
+
+  assert(message);
+
+  message_length = strlen(message);
+  if (message_length == 0) {
+    perror("Message to send to $NOTIFY_SOCKET is empty");
+    return -EINVAL;
+  }
+
+  socket_path = getenv("NOTIFY_SOCKET");
+  if (!socket_path)
+    return 0; /* Not running under systemd? Nothing to do */
+
+  path_length = strlen(socket_path);
+  /* Ensure there is room for NULL byte */
+  if (path_length >= sizeof(socket_addr.sun_path)) {
+    perror("$NOTIFY_SOCKET contains a path that is too long for AF_UNIX");
+    return -E2BIG;
+  }
```

Comment:
As per docs, `@` is used to denote abstract namespace sockets. Please implement this too, i.e. add:

```c
if (socket.addr.sun_path[0] == '@') /* Handle abstract namespace sockets */
        socket.addr.sun_path[0] = 0;
```

also, ad per docs, only if first char is `/` it's an AF_UNIX socket in the fs. Hence better refuse all other cases. This matters, because we nowadays allow AF_VSOCK too, for VM ready notification. (I don't think we should implement AF_VSOCK here, only systemd itself really should need that, but I still think we should correctly fail in that case, with _EAFNOSUPPORT or so).